developer_guides/virtualisation/virtualisation.rst

```diff
@@ -0,0 +1,255 @@
+
+3.2. Streaming
+--------------
+
+Guest audio streaming is mostly transparent for the host. The host audio
+subsystem doesn't get involved with most guest streaming or kcontrol operations.
+All those operations are completely processed on the guest and only resulting
+DSP IPCs are forwarded to the host. The only exceptions are beginning and ending
+of the streaming, at which times the ASoC subsystem on the host has to be
+informed, that the respective virtual pipeline, typically including a physical
+codec DAI, is becoming active or inactive respectively. Deactivation of the
+guest pipeline on the host requires changes to the ASoC core, without which the
+guest audio path is not recognised as inactive and is not pruned, thus making
+further streaming by the guest impossible.
+
+Upon reception of the STREAM_TRIG_START IPC message the vhost driver activates
+the virtual PCM pipeline and updates the routing information on the host.
+
+As mentioned above, it is our goal to implement data zero-copy. As long as
+this isn't the case, the guest VirtIO driver implements .block_write() and
+.block_read() SOF DSP operation methods for data streaming via the data
+VirtQueue.
+
+The host receives requests on the data VirtQueue and copies data between PCM
+runtime buffers and the queue and responds either with a status or with a data
+buffer.
+
+Similarly to the data, streaming buffer position updates are currently also
+transferred via a dedicated VirtQueue. Both the host and the guest are
+configured to use IPC messages for position updates. Buffers for VirtIO
+VirtQueues are always provided by guests, therefore there should always be a
+position update buffer waiting on the host side to send a position update
+message as soon as one arrives from the DSP.
+
+Upon reception of the STREAM_TRIG_STOP IPC message the vhost driver updates
+the routing information and deactivates the virtual PCM pipeline on the host.
+This operation requires particular care, because in the present state, just
+calling
+
+	soc_dpcm_runtime_update();
+	snd_soc_runtime_deactivate();
+
+doesn't deactivate the pipeline and therefore doesn't allow runtime PM to
+suspend the interface, which then breaks following activation attempts. The
+current VirtIO implementation contains a fix for that, but that fix might turn
+out to be incorrect. The difficulty in establishing a correct fix is, that so
+far it hasn't been possible to find a test case for
+snd_soc_runtime_deactivate().
+
+3.3. Deactivation
+-----------------
+
+When the guest isn't actively using audio, we want to allow the host to
+runtime suspend the DSP. That usually means switching the DSP off. Next time the
+guest wants to activate audio it will have to perform initialisation again,
+including sending its topology to the host. This happens automatically as long
+as the guest runtime suspends too or performs the initialisation cycle in some
+other way, by unloading and reloading drivers, or by rebooting.
+
```

Comment:
guest should cache it's topology so it does not need to ask for it from host at PM resume.

Reply:
@lgirdwood It does cache it of course. It resends it to the host - the host doesn't cache topology fragments of all guests, don't think that would make sense. And, of course, when the guest reboots, or reloads the driver, it requests the topology again.

Reply:
The host should at least have an idea of the pipelines and components that belong to each guest. This way it can delete them in FW if guest crashes etc.

Reply:
@lgirdwood the host has guest component ID ranges, is that enough?

Reply:
yes, but it needs to tell FW to delete the pipeline and components if needed. host may need to keep a list of all guest comps/pipes for each guest.